### ISSUE TITLE
Reduce the size of the repository

### ISSUE THREAD

**Issue by i-give-up** [marius-sucan/KeyPress-OSD#5]
Disclaimer: I'm still a beginner in Git/GitHub so what I say below may have inaccuracies. Please take backups and/or test with test repositories in case something goes wrong.

I cloned this repository using `git clone https://github.com/marius-sucan/KeyPress-OSD.git` and found the folder size to be about 110MB. This is surprising because the files themselves (including the zip) is just about 10MB.

It seems that the .git folder is 100MB. After some digging around, this seems to be due to zip/exe files being committed to the repository which caused the repository's size to balloon. If I'm understanding correctly, for text files e.g. scripts and source code, Git can just store the difference in text whereas for binary files (zip/exe) Git may have to store the entire file every time a new version is committed.

Using the steps in [this blogpost](http://stevelorek.com/how-to-shrink-a-git-repository.html), I managed to shrink the repository size to around 3MB. It removed the "keypress-osd-compiled.zip" though. The steps I used to do this

1. In the command line, cd to the project directory
1. `git filter-branch --tag-name-filter cat --index-filter 'git rm -r --cached --ignore-unmatch *.exe *.zip' --prune-empty -f -- --all`
1. Delete the directory ".git/refs/original"
1. `git reflog expire --expire=now --all`
1. `git gc --aggressive --prune=now`

These steps I haven't done yet due to me not having a repository. I'm planning to test them later on a fork though.

1. `git push origin --force --all`
1. `git push origin --force --tags`

For compiled files or zips containing compiled files, the standard practice seems to be adding them to Releases instead of committing them to the repository. For example, [AutoHotkey_L's Release page](https://github.com/Lexikos/AutoHotkey_L/releases) has the `AutoHotkey_<version>_setup.exe` files but those setup.exe's are not in [the repository itself](https://github.com/Lexikos/AutoHotkey_L). Other examples: [wincompose](https://github.com/samhocevar/wincompose/releases), [greenshot](https://github.com/greenshot/greenshot/releases), [pandoc](https://github.com/jgm/pandoc/releases).

One thing that may be affected is the Check Update feature within the script but I'm not sure how it is implemented.

**Comment by i-give-up** [comment]
So I tested with my own fork: cloned a local copy, shrank repository as described above, then ran `git push` described above. After that, I deleted the local copy and cloned a fresh copy of the fork.

I tried executing `keypress-osd.ahk` on a few randomly picked versions using AutoHotkey_H and they worked fine except for a very old version v2.50.

I also tried compiling some versions to .exe and noticed that `updater.bat` which is needed for compilation is not in the repository. It can only be found in the `keypress-osd-compiled.zip`. I also found `msvcr100.dll` in the zip file though that file doesn't seem to be needed for compilation.

To summarize, removing the zip/exe files from the repository and Git index as described above shouldn't cause issue in terms of executing the .ahk file or compiling it into .exe except that the required `updater.bat` is not in the repository. Also, I haven't looked into how it would affect Check Update feature.

**Comment by marius-sucan** [comment]
Hello!
I am very new to GitHub as well, and as far as I can see.... below your skills with it.

The updater downloads all the files from my own server. The BAT file is used only for the compiled edition, to replace itself.

To compile it, one needs everything inside Lib\ and updater.bat. The compiled file extracts its own files automatically, eg. the help files. The other scripts and the sound files are embedded as well, but used as internal libs - inside the EXE file. Msvcr100.dll is used by AHK itself.

The reason I have the ZIP file here, is to ensure an easy access to a package containing AHK-H along with KeyPress OSD source files and KeyPress OSD compiled, for both x32 and x64. I noticed the role of Releases on GitHub, but I do not know how one can better use this feature.

Can I push to 'Releases' the ZIP file and not in the repository, to avoid making it huge?

Thank you.

Best regards, Marius.

**Comment by i-give-up** [comment]
> Can I push to 'Releases' the ZIP file and not in the repository, to avoid making it huge?

Yes. It's step 7 in [this guide](https://help.github.com/articles/creating-releases/). I think you can only do it via the web interface though. There's an API for this but I'm not familiar with it. 

For existing releases, you can add binary files by going to [the Tags page](https://github.com/marius-sucan/KeyPress-OSD/tags), then clicking "Add Release Notes". There's "Attach binaries by dropping them here or selecting them".

You also have to perform the steps in the first post (including the `git push` steps) to remove the files from the Git index because that's the main reason for the repository size being large.

**Comment by marius-sucan** [comment]
Keypress-osd.exe it is just AHK-H binary renamed. It automatically executes the script file with the same name. The -compiled.exe is the actually compiled script.

Okay, I will do what you suggested me, but not very soon. I still have a myriad of bugs lurking into the script and I am struggling to get them fixed.

Thank you !

Best regards, Marius.

**Comment by marius-sucan** [comment]
Hello, again!

Have you tested the commands to prune the EXEs/ZIPs from the repository? I want to reduce its size as you suggested. Thank you for your help.

Best regards, Marius.

**Comment by marius-sucan** [comment]
I finally managed to do it. I removed all the ZIPs with the compiled script. The repository is now down to 30 mb from 450 megabytes :-).

Thank you very much.

Best regards, Marius.

**Comment by i-give-up** [comment]
Hm... I don't know what happened here. 

When I posted the first message in this issue, the repository's original size is ~110 MB and reduced size is ~3MB.

But I just tried cloning the repository and it's more than 130 MB now.

I'm not proficient in Git enough to know what's going on.

**Comment by marius-sucan** [comment]
Hello!

Thanks for getting back to this.

Since then [in April 2018, when you posted the suggestion], I made many pushes to GitHub which contained the ZIP file, until in June 2018, so that would explain why I reported a larger size.

The sizes I reported were based on my local folder size.

The clone you made, did it contain any ZIP file of the compiled KeyPress OSD ? I do not know why here it is only 30 megabytes.

Best regards, Marius.

**Comment by i-give-up** [comment]
There's no zip file in the cloned directory. Yeah, I have no idea what's going on either.
